File: sniffer.py
import os
import json
import firebase_admin
from firebase_admin import credentials
from google.cloud import firestore
from google.oauth2 import service_account

print("--- SNIFFER.PY EXECUTION START ---")

# --- Firebase Initialization at the very top ---
PROJECT_ID = "gen-lang-client-0472035720"
DATABASE_ID = "ai-studio-edf518e7-ccd7-4d8a-afd7-3f1030781b80"

if not firebase_admin._apps:
    try:
        if os.environ.get('FIREBASE_SERVICE_ACCOUNT'):
            service_account_info = json.loads(os.environ.get('FIREBASE_SERVICE_ACCOUNT'))
            cred = credentials.Certificate(service_account_info)
            firebase_admin.initialize_app(cred, {
                'projectId': PROJECT_ID
            })
            print(f"[Firebase] Initialized with service account from environment (Project: {PROJECT_ID}).")
            
            # Create explicit credentials for the direct Firestore client
            google_creds = service_account.Credentials.from_service_account_info(service_account_info)
            db = firestore.Client(project=PROJECT_ID, database=DATABASE_ID, credentials=google_creds)
        elif os.path.exists('serviceAccountKey.json'):
            cred = credentials.Certificate('serviceAccountKey.json')
            firebase_admin.initialize_app(cred)
            print("[Firebase] Initialized with serviceAccountKey.json.")
            
            # Create explicit credentials for the direct Firestore client
            google_creds = service_account.Credentials.from_service_account_file('serviceAccountKey.json')
            db = firestore.Client(project=PROJECT_ID, database=DATABASE_ID, credentials=google_creds)
        else:
            config_path = 'firebase-applet-config.json'
            if os.path.exists(config_path):
                with open(config_path, 'r') as f:
                    config = json.load(f)
                    firebase_admin.initialize_app(options={'projectId': config.get('projectId', PROJECT_ID)})
                    print(f"[Firebase] Initialized with project ID from config: {config.get('projectId')}")
            else:
                firebase_admin.initialize_app()
                print("[Firebase] Initialized with default credentials.")
            db = firestore.Client(project=PROJECT_ID, database=DATABASE_ID)
        print('Firebase Initialized')
    except Exception as e:
        print(f"Firebase initialization fatal error: {e}")
        raise e

import time
import re
import logging
from datetime import datetime
import requests
from bs4 import BeautifulSoup
import gc

# --- Configuration ---

logging.basicConfig(
    level=logging.INFO, 
    format='%(asctime)s - %(levelname)s - %(message)s',
    handlers=[
        logging.StreamHandler(),
        logging.FileHandler("sniffer.log")
    ]
)

HEADERS = {
    "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/119.0.0.0 Safari/537.36",
    "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8",
    "Accept-Language": "en-US,en;q=0.9",
}

def log_to_firestore(message, level='info', device_id=None):
    try:
        db.collection('logs').add({
            'timestamp': firestore.SERVER_TIMESTAMP,
            'level': level,
            'message': message,
            'deviceId': device_id
        })
    except Exception as e:
        logging.error(f"Failed to log to Firestore: {e}")

def scrape_receive_smss():
    """Scrapes receive-smss.com for new numbers."""
    try:
        logging.info("Scraping receive-smss.com...")
        response = requests.get("https://receive-smss.com/", headers=HEADERS, timeout=30)
        logging.info(f"Status Code (receive-smss.com): {response.status_code}")
        log_to_firestore(f"Status Code (receive-smss.com): {response.status_code}")
        
        # Heartbeat Check: Log first 2000 chars
        log_to_firestore(f"Heartbeat Check (2000 chars): {response.text[:2000]}")
        
        if "Cloudflare" in response.text or "Access Denied" in response.text:
            logging.error("BLOCKED by Cloudflare/Access Denied on receive-smss.com")
            log_to_firestore("BLOCKED by Cloudflare on receive-smss.com", level='error')
        else:
            log_to_firestore("Heartbeat: No Cloudflare detected.", level='info')
            
        response.raise_for_status()
        soup = BeautifulSoup(response.text, 'html.parser')
        numbers_data = []
        cards = soup.select(".number-boxes-item")
        logging.info(f"Found {len(cards)} cards on receive-smss.com")
        log_to_firestore(f"Found {len(cards)} cards on receive-smss.com")
        for card in cards:
            try:
                num_elem = card.select_one(".number-boxes-item-number")
                if not num_elem: continue
                clean_number = re.sub(r'\D', '', num_elem.get_text(strip=True))
                if "New" in card.get_text():
                    numbers_data.append({
                        'number': clean_number,
                        'source': 'receive-smss.com',
                        'status': 'found',
                        'addedAt': firestore.SERVER_TIMESTAMP
                    })
            except: pass
        del soup
        gc.collect()
        return numbers_data
    except Exception as e:
        logging.error(f"Error scraping receive-smss: {e}")
        return []

def scrape_receive_sms_free():
    """Scrapes receive-sms-free.cc for new numbers."""
    try:
        logging.info("Scraping receive-sms-free.cc...")
        response = requests.get("https://receive-sms-free.cc/", headers=HEADERS, timeout=30)
        response.raise_for_status()
        soup = BeautifulSoup(response.text, 'html.parser')
        numbers_data = []
        # Example selector, might need adjustment
        links = soup.select("a[href*='/free-sms-number/']")
        for link in links:
            try:
                text = link.get_text(strip=True)
                clean_number = re.sub(r'\D', '', text)
                if len(clean_number) > 8:
                    numbers_data.append({
                        'number': clean_number,
                        'source': 'receive-sms-free.cc',
                        'status': 'found',
                        'addedAt': firestore.SERVER_TIMESTAMP
                    })
            except: pass
        del soup
        gc.collect()
        return numbers_data
    except Exception as e:
        logging.error(f"Error scraping receive-sms-free: {e}")
        return []

def scrape_mobilesms_free():
    """Scrapes mobilesms.io/free for new numbers."""
    try:
        logging.info("Scraping mobilesms.io/free...")
        response = requests.get("https://mobilesms.io/free-sms-numbers/", headers=HEADERS, timeout=30)
        response.raise_for_status()
        soup = BeautifulSoup(response.text, 'html.parser')
        numbers_data = []
        # Example selector
        items = soup.select(".free-number-item")
        for item in items:
            try:
                num_text = item.get_text(strip=True)
                clean_number = re.sub(r'\D', '', num_text)
                if len(clean_number) > 8:
                    numbers_data.append({
                        'number': clean_number,
                        'source': 'mobilesms.io',
                        'status': 'found',
                        'addedAt': firestore.SERVER_TIMESTAMP
                    })
            except: pass
        del soup
        gc.collect()
        return numbers_data
    except Exception as e:
        logging.error(f"Error scraping mobilesms: {e}")
        return []

def monitor_otp(number_doc_id, phone_number, source):
    """Monitors the inbox of a specific number for a WhatsApp OTP."""
    url = ""
    if source == 'receive-smss.com':
        url = f"https://receive-smss.com/sms/{phone_number}/"
    elif source == 'receive-sms-free.cc':
        url = f"https://receive-sms-free.cc/free-sms-number/{phone_number}.html"
    elif source == 'mobilesms.io':
        url = f"https://mobilesms.io/free-sms-numbers/{phone_number}/"
    
    if not url: return False

    logging.info(f"Monitoring OTP for {phone_number} at {url}")
    
    start_time = time.time()
    timeout = 600 # 10 minutes timeout
    
    while time.time() - start_time < timeout:
        try:
            # Check if we should still monitor (e.g. status changed to 'requesting_otp' by Node.js)
            doc = db.collection('numbers').document(number_doc_id).get()
            if not doc.exists: break
            status = doc.to_dict().get('status')
            
            # Only monitor if Node.js has triggered the OTP request
            if status == 'requesting_otp':
                response = requests.get(url, headers=HEADERS, timeout=30)
                response.raise_for_status()
                soup = BeautifulSoup(response.text, 'html.parser')
                text = soup.get_text()
                
                # Look for WhatsApp OTP
                otp_match = re.search(r'WhatsApp(?: code)?:?\s*(\d{3}-?\d{3})', text, re.IGNORECASE)
                if otp_match:
                    otp = otp_match.group(1).replace('-', '')
                    logging.info(f"OTP Found for {phone_number}: {otp}")
                    db.collection('numbers').document(number_doc_id).update({
                        'otp': otp,
                        'status': 'otp_found'
                    })
                    del soup
                    gc.collect()
                    return True
                del soup
                gc.collect()
            
            time.sleep(15)
        except Exception as e:
            logging.error(f"Error monitoring {phone_number}: {e}")
            time.sleep(10)
            
    return False

def main():
    logging.info("Scavenger Farm Engine Started")
    log_to_firestore("Scavenger Farm Engine Started")
    
    # Force a Save: Test Number to confirm DB connection
    try:
        test_num = {
            'number': 'TEST_NUMBER_12345',
            'source': 'heartbeat_check',
            'status': 'found',
            'addedAt': firestore.SERVER_TIMESTAMP
        }
        db.collection('numbers').add(test_num)
        logging.info("Force Save Successful: Test Number added to Firestore.")
        log_to_firestore("Force Save Successful: Test Number added to Firestore.")
    except Exception as e:
        logging.error(f"Force Save FAILED: {e}")

    while True:
        try:
            all_found = []
            all_found.extend(scrape_receive_smss())
            all_found.extend(scrape_receive_sms_free())
            all_found.extend(scrape_mobilesms_free())
            
            for num_data in all_found:
                # Check if number already exists
                existing = db.collection('numbers').where('number', '==', num_data['number']).limit(1).get()
                
                if not existing:
                    doc_ref = db.collection('numbers').add(num_data)[1]
                    logging.info(f"New Scavenged Number: {num_data['number']}")
                    log_to_firestore(f"New Scavenged Number: {num_data['number']}")
                    
                    # Start a thread or just monitor in sequence (for now sequence is simpler)
                    # In a real farm we'd use threading, but let's keep it simple for now
                    monitor_otp(doc_ref.id, num_data['number'], num_data['source'])
            
            logging.info("Cycle complete. Sleeping 10 minutes...")
            gc.collect()
            time.sleep(600)
        except Exception as e:
            logging.error(f"Main loop error: {e}")
            time.sleep(60)

if __name__ == "__main__":
    main()

File: sniffer.ts
import { chromium } from 'playwright-extra';
import stealth from 'playwright-stealth';
import { Browser, Page } from 'playwright';
import admin from 'firebase-admin';
import { getFirestore, FieldValue } from 'firebase-admin/firestore';
import axios from 'axios';

// Use stealth plugin
chromium.use(stealth());

const API_URL = process.env.APP_URL || 'http://localhost:3000';

// --- Firebase Initialization ---
if (!admin.apps.length) {
  admin.initializeApp();
}

const db = getFirestore();
const GHOST_LIMIT = 5;

async function logToFirestore(message: string, level: string = 'info', deviceId?: string) {
  try {
    await db.collection('logs').add({
      timestamp: FieldValue.serverTimestamp(),
      level,
      message,
      deviceId: deviceId || null
    });
    console.log(`[${level.toUpperCase()}] ${message}`);
  } catch (err) {
    console.error("Failed to log to Firestore:", err);
  }
}

async function getActiveGhostCount() {
  const snapshot = await db.collection('numbers').where('status', '==', 'active').get();
  return snapshot.size;
}

async function setupPage(browser: Browser): Promise<Page> {
  const context = await browser.newContext({
    userAgent: 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/119.0.0.0 Safari/537.36'
  });
  
  const page = await context.newPage();
  
  // Apply Stealth
  // @ts-ignore
  // await stealth(page); // stealth is usually applied to context or browser in some versions, but playwright-stealth is often used like this:
  
  // Resource Blocking for Ultra-Light Memory Usage
  await page.route('**/*', (route) => {
    const type = route.request().resourceType();
    if (['image', 'stylesheet', 'font', 'media'].includes(type)) {
      route.abort();
    } else {
      route.continue();
    }
  });

  return page;
}

async function scrapeReceiveSmss(page: Page) {
  try {
    await logToFirestore("Scraping receive-smss.com with Playwright...");
    await page.goto("https://receive-smss.com/", { waitUntil: 'domcontentloaded', timeout: 60000 });
    
    const numbers = await page.evaluate(() => {
      const cards = Array.from(document.querySelectorAll(".number-boxes-item"));
      return cards.map(card => {
        const numElem = card.querySelector(".number-boxes-item-number");
        const countryElem = card.querySelector(".number-boxes-item-country");
        const linkElem = card.querySelector("a");
        return {
          number: numElem?.textContent?.replace(/\s+/g, '').trim(),
          country: countryElem?.textContent?.trim(),
          url: linkElem?.href,
          source: 'receive-smss.com'
        };
      }).filter(n => n.number);
    });

    return numbers;
  } catch (err: any) {
    await logToFirestore(`Error scraping receive-smss: ${err.message}`, "error");
    return [];
  }
}

async function scrapeTextVerified(page: Page) {
  try {
    await logToFirestore("Scraping textverified.com/free with Playwright...");
    await page.goto("https://www.textverified.com/free", { waitUntil: 'domcontentloaded', timeout: 60000 });
    
    const numbers = await page.evaluate(() => {
      const cards = Array.from(document.querySelectorAll(".free-number-card"));
      return cards.map(card => {
        const numElem = card.querySelector(".number");
        const countryElem = card.querySelector(".country");
        const linkElem = card.querySelector("a");
        return {
          number: numElem?.textContent?.replace(/\s+/g, '').trim(),
          country: countryElem?.textContent?.trim(),
          url: linkElem?.href,
          source: 'textverified.com'
        };
      }).filter(n => n.number);
    });

    return numbers;
  } catch (err: any) {
    await logToFirestore(`Error scraping textverified: ${err.message}`, "error");
    return [];
  }
}

async function scrapeVeepn(page: Page) {
  try {
    await logToFirestore("Scraping veepn.com/online-sms with Playwright...");
    await page.goto("https://veepn.com/online-sms/", { waitUntil: 'domcontentloaded', timeout: 60000 });
    
    const numbers = await page.evaluate(() => {
      // Assuming structure based on common patterns if not specified
      const links = Array.from(document.querySelectorAll("a[href*='/online-sms/number-']"));
      return links.map(link => {
        const text = link.textContent?.trim() || "";
        return {
          number: text.includes("+") ? text.replace(/\s+/g, '') : null,
          country: 'Global',
          url: (link as HTMLAnchorElement).href,
          source: 'veepn.com'
        };
      }).filter(n => n.number);
    });

    return numbers;
  } catch (err: any) {
    await logToFirestore(`Error scraping veepn: ${err.message}`, "error");
    return [];
  }
}

async function monitorActivation(page: Page, docId: string, url: string) {
  await logToFirestore(`Monitoring activation for ${docId} at ${url}`);
  try {
    await page.goto(url, { waitUntil: 'domcontentloaded', timeout: 60000 });
    
    // Refresh loop
    for (let i = 0; i < 30; i++) { // Max 5 minutes (30 * 10s)
      await page.reload({ waitUntil: 'domcontentloaded' });
      const content = await page.content();
      
      // Look for WhatsApp and 6-digit code
      const match = content.match(/WhatsApp.*?(\d{6})/i) || content.match(/(\d{6}).*?WhatsApp/i);
      
      if (match) {
        const code = match[1];
        await logToFirestore(`FOUND WHATSAPP CODE: ${code} for ${docId}`);
        
        // Update Firestore
        await db.collection('numbers').doc(docId).update({
          otpCode: code,
          status: 'code_received',
          updatedAt: FieldValue.serverTimestamp()
        });

        // Notify Backend
        try {
          await axios.post(`${API_URL}/api/sniffer/code`, { docId, code });
        } catch (err: any) {
          console.error(`Failed to notify backend: ${err.message}`);
        }
        
        return true;
      }
      
      await new Promise(resolve => setTimeout(resolve, 10000));
      
      // Check if status changed (e.g. cancelled)
      const doc = await db.collection('numbers').doc(docId).get();
      if (doc.data()?.status !== 'activating') {
        await logToFirestore(`Activation cancelled for ${docId}`);
        return false;
      }
    }
    
    await db.collection('numbers').doc(docId).update({
      status: 'timeout',
      updatedAt: FieldValue.serverTimestamp()
    });
    await logToFirestore(`Activation timeout for ${docId}`);
  } catch (err: any) {
    await logToFirestore(`Error monitoring activation: ${err.message}`, "error");
  }
  return false;
}

async function saveNumbers(numbers: any[]) {
  for (const num of numbers) {
    try {
      const q = await db.collection('numbers').where('number', '==', num.number).get();
      if (q.empty) {
        await db.collection('numbers').add({
          ...num,
          status: 'found',
          addedAt: FieldValue.serverTimestamp()
        });
        await logToFirestore(`New number found: ${num.number} (${num.source})`);
      }
    } catch (err: any) {
      console.error(`Error saving number ${num.number}:`, err.message);
    }
  }
}

async function main() {
  await logToFirestore("Stealth-Light Scavenger Engine Started");

  const browser = await chromium.launch({
    args: ['--disable-setuid-sandbox', '--no-sandbox', '--disable-dev-shm-usage', '--single-process']
  });

  try {
    while (true) {
      // 1. Check for Activating Numbers
      const activatingSnapshot = await db.collection('numbers').where('status', '==', 'activating').limit(1).get();
      if (!activatingSnapshot.empty) {
        const doc = activatingSnapshot.docs[0];
        const data = doc.data();
        if (data.url) {
          const page = await setupPage(browser);
          await monitorActivation(page, doc.id, data.url);
          await page.close();
        }
      }

      // 2. Check Ghost Limit
      const activeCount = await getActiveGhostCount();
      if (activeCount >= GHOST_LIMIT) {
        await logToFirestore(`Ghost Limit Reached (${activeCount}/${GHOST_LIMIT}). Scavenger sleeping...`);
        await new Promise(resolve => setTimeout(resolve, 60000));
        continue;
      }

      // 3. Scrape for New Numbers
      const page = await setupPage(browser);
      const s1 = await scrapeReceiveSmss(page);
      const s2 = await scrapeTextVerified(page);
      const s3 = await scrapeVeepn(page);
      await page.close();

      const allNumbers = [...s1, ...s2, ...s3];
      if (allNumbers.length > 0) {
        await saveNumbers(allNumbers);
      }

      console.log("Main loop cycle complete. Sleeping for 60s...");
      await new Promise(resolve => setTimeout(resolve, 60000));
    }
  } catch (err: any) {
    await logToFirestore(`CRITICAL ENGINE FAILURE: ${err.message}`, "error");
  } finally {
    await browser.close();
  }
}

main().catch(async err => {
  console.error("Fatal error in main:", err);
  process.exit(1);
});
